kubefed2 disable should delete the FTC rather than set propagationEnabled

font opened the pull request:
With this change `kubefed2 disable` now deletes the `FederatedTypeConfig` instead of just setting `propagationEnabled` to `false`. It will also wait for the sync controller to shut down before deleting it, and optionally delete the federated type CRD as well.

Partial fix for #721.

xunpan commented on pkg/kubefed2/disable.go:
```diff
@@ -37,11 +37,17 @@ import (
 	fedv1a1 "github.com/kubernetes-sigs/federation-v2/pkg/apis/core/v1alpha1"
 	genericclient "github.com/kubernetes-sigs/federation-v2/pkg/client/generic"
 	ctlutil "github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+
```
Is this convention to add a space line here? It seems not necessary to me.

font replied:
Not a convention, just an artifact left over from a previous `goimports` issue. I've removed.

xunpan commented on pkg/kubefed2/disable.go:
```diff
@@ -252,6 +284,15 @@ func deleteFederatedTypeConfig(client genericclient.Client, typeConfig *fedv1a1.
+	if err != nil {
+		return nil, err
+	}
+	typeConfig := enable.GenerateTypeConfigForTarget(*apiResource, enableTypeDirective).(*fedv1a1.FederatedTypeConfig)
```
This works. But as caller functions only needs `typeconfig.Interface`, it does not needs such type conversion here.

font replied:
This type assertion is necessary because this function returns a `*fedv1a1.FederatedTypeConfig`. It returns this because the `DisableFederation` caller is re-using its `typeConfig` variable (for simplicity) that is defined as a `*fedv1a1.FederatedTypeConfig` [here](https://github.com/kubernetes-sigs/federation-v2/pull/769/files#diff-fc79e5fb1808a541a91f4994f5f02254R187). `typeConfig` is defined this way because it is passed to multiple functions that expect different interface implementations e.g. `runtime.Object`, `typeconfig.Interface`.

xunpan replied:
I'm fine with the code anyway. 
I guess it is compromise for 2 case:
 - use a common variable in caller function and has type conversion in the function
 - use a different temparary variable in caller function and has no type conversion internally
Outter logic makes more sense than inner logic for code reading.

font replied:
It is a compromise as you say. I prefer this implementation for a couple reasons:
1. It provides the most flexibility to the caller because it returns the same type passed in so that the caller is free to choose what interface implementation to use without performing any type assertion.
2. If I used a temporary variable in `DisableFederation`, I'd have to make 2 different calls to `deleteFederatedType` using 2 different `typeConfig` variables. Instead, this approach simplifies the code in `DisableFederation`.

I am happy to change it if you and others feel strongly.

irfanurrehman replied:
I think its fine the way it is. Type conversion is anyways needed at some place, because of what @font mentioned [here](https://github.com/kubernetes-sigs/federation-v2/pull/769/commits/2f9112de2089e8f6b58415c7686b91df784d6cc8#r276788968). Also I think it makes sense to have this function return the concrete type as the caller can receive it in either the concrete type or an interface.
Also, let this not lead to bikeshedding.. :)

irfanurrehman commented on pkg/kubefed2/disable.go:
```diff
@@ -148,48 +151,104 @@ func DisableFederation(cmdOut io.Writer, config *rest.Config, typeConfigName ctl
 		return errors.Wrap(err, "Failed to get federation clientset")
 	}
+
+	write := func(data string) {
```
[Not necessary to be done in this PR] This could be a common utility function both `enable` and `disable` could use.

irfanurrehman commented on pkg/kubefed2/disable.go:
```diff
@@ -148,48 +151,104 @@ func DisableFederation(cmdOut io.Writer, config *rest.Config, typeConfigName ctl
+	return false, errors.Wrapf(err, "Error retrieving FederatedTypeConfig %q", typeConfigName)
+}
+
+func disablePropagationController(client genericclient.Client, typeConfig *fedv1a1.FederatedTypeConfig, typeConfigName ctlutil.QualifiedName, write func(string)) error {
```
Isn't this `stopPropagationController`, meaning the controller for this type would actually be stopped; right?

irfanurrehman commented on pkg/kubefed2/disable.go:
```diff
@@ -104,17 +110,35 @@ func NewCmdTypeDisable(cmdOut io.Writer, config util.FedConfig) *cobra.Command {
-	if len(args) == 0 {
-		return errors.New("NAME is required")
+	j.enableTypeDirective = enable.NewEnableTypeDirective()
+	fd := j.enableTypeDirective
```
`fd` is probably confusing, `etd` or just `td` would be better for a reader with no prior code knowledge.. :-)

irfanurrehman commented:
@font apologies for delayed look at this. Few naming nits, otherwise good to go.

font commented on pkg/kubefed2/disable.go:
```diff
@@ -139,60 +166,132 @@ func (j *disableType) Run(cmdOut io.Writer, config util.FedConfig) error {
+	err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+		err := client.Get(context.TODO(), typeConfig, typeConfigName.Namespace, typeConfigName.Name)
+		if err != nil {
+			return false, err
```
We can timeout instead. I've updated.

font commented on pkg/kubefed2/disable.go:
```diff
@@ -139,60 +166,132 @@ func (j *disableType) Run(cmdOut io.Writer, config util.FedConfig) error {
-	err = deleteFederatedType(config, typeConfig, write)
 	if err != nil {
-		return err
+		write(fmt.Sprintf("WARNING: unable to verify propagation controller for FederatedTypeConfig %q is stopped: %v\n", typeConfigName, err))
```
It is intentional. I don't think `check` is as strong as something like `verify`, which would suggest making sure something is done. I'm open to other suggestions for names.

marun replied:
Why is it desirable to wait 10s for the controller to stop but not fail if it takes longer than that?  Is that an indication that disabling propagation is unnecessary and the ftc can just be deleted without first being disabled?

font commented on pkg/kubefed2/disable.go:
```diff
@@ -139,60 +166,132 @@ func (j *disableType) Run(cmdOut io.Writer, config util.FedConfig) error {
 	}
-	write(fmt.Sprintf("federatedtypeconfig %q deleted\n", typeConfigName))
-
+	write(fmt.Sprintf("FederatedTypeConfig %q deleted\n", typeConfigName))
```
Ah, I was wondering about that as it seemed inconsistent with the rest of the logs using camel case.

font commented on pkg/kubefed2/disable.go:
```diff
@@ -202,8 +301,13 @@ func deleteFederatedType(config *rest.Config, typeConfig typeconfig.Interface, w
 	err = client.CustomResourceDefinitions().Delete(crdName, nil)
 	if err != nil {
-		return errors.Wrapf(err, "Error deleting crd %q", crdName)
+		if apierrors.IsNotFound(err) {
```
Thanks, simplified to something similar.

font commented on pkg/kubefed2/disable.go:
```diff
@@ -202,8 +301,13 @@ func deleteFederatedType(config *rest.Config, typeConfig typeconfig.Interface, w
 	crdName := typeconfig.GroupQualifiedName(typeConfig.GetFederatedType())
 	err = client.CustomResourceDefinitions().Delete(crdName, nil)
```
The latter is not implemented in this PR due to the following per our offline conversation:

1. Needing to discover the federation config (it has since been merged).
2. Possibly resolving it with docs instead of code due to member cluster resources.

With that in mind, are you suggesting we implement this in code before merging this or that we should at least add a `--force` option prior to merging?

font commented:
Thanks @irfanurrehman and @marun for the review! I've pushed a new commit addressing most comments and will need to discuss a couple more.

font commented:
@irfanurrehman @marun I've rebased and added a check to avoid deleting the federated type CRD if any of its CRs exist. Pending your review, I think that should allow this to merge and we can think about how to support a possible `--force` option to also remove the CRs prior to deleting the CRD while considering cluster and namespaced scoped deployments.

irfanurrehman commented on pkg/kubefed2/disable.go:
```diff
@@ -298,12 +299,45 @@ func generatedFederatedTypeConfig(config *rest.Config, enableTypeDirective *enab
+	if err != nil {
+		return err
+	} else if exists {
+		return errors.Errorf("Cannot delete CRD %q while resource instances exist. Please delete them and try again or remove --delete-crd option to disable only\n", crdName)
```
I found the language a little confusing.
Consider rewording to "Please try `kubefed2 disable` again after removing the resource instances or without the option `--delete-CRD`"
Or something similar. The usage of the word disable is a little tricky, I think.

font replied:
Updated, PTAL.

irfanurrehman commented:
> @irfanurrehman @marun I've rebased and added a check to avoid deleting the federated type CRD if any of its CRs exist. Pending your review, I think that should allow this to merge and we can think about how to support a possible `--force` option to also remove the CRs prior to deleting the CRD while considering cluster and namespaced scoped deployments.

+1 to implementing `--force` in a follow up.

irfanurrehman commented:
Thanks @font. lgtm. Would wait for @marun to verify (and merge).

marun commented on pkg/kubefed2/disable.go:
```diff
@@ -139,75 +167,184 @@ func (j *disableType) Run(cmdOut io.Writer, config util.FedConfig) error {
+	err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+		typeConfig = &fedv1a1.FederatedTypeConfig{}
+		err := client.Get(context.TODO(), typeConfig, typeConfigName.Namespace, typeConfigName.Name)
+		if err != nil {
```
Is there a good reason not to log this error?

font replied:
Avoiding dumping 100 errors to the console buffer in the event the error persists for the entire duration of the poll. I went ahead and added it since it's not in the normal path we expect.

marun replied:
(No action required) Hmm, maybe we need a wait loop that does exponential backoff on error?

marun commented on pkg/kubefed2/disable.go:
```diff
@@ -139,75 +167,184 @@ func (j *disableType) Run(cmdOut io.Writer, config util.FedConfig) error {
+				return err
+			}
 		}
+		err = deleteFederatedType(config, typeConfig, write)
```
I'm recalling the original reason why the propagation was being disabled.  Deleting the crd while propagation is still being performed is likely to generate a bunch of spurious error messages in the controller manager log as the watches are terminated and can't be reestablished.  Deleting the FTC before the CRD risks breaking idempotent behavior if deletion of the CRD is subsequently unsuccessful, since a subsequent disable attempt may not be able to find the CRD again (if it has non-standard group/version/name/etc).  Disabling, deleting CRD, then deleting FTC would seem the cleanest path.

font replied:
I see, makes sense. Updated.

font commented:
@marun I've rebased and updated based on your comments. PTAL.

marun commented:
@font LGTM.   

@irfanurrehman  Over to you for a final review.

irfanurrehman commented:
@font the CI failure looks like a flake but seems legit. Not sure if increasing the timeout would help anything. 10 seconds seems to be a good enough time for the  controller to be stopped.

font commented:
Thanks @irfanurrehman! I've addressed the e2e issue and this is now ready for your re-review.

irfanurrehman commented:
@font thanks for doing this.
/lgtm
